integ/src/monitor.rs

```diff
@@ -0,0 +1,533 @@
+use kube::api::{Api, ListParams, ObjectList};
+use snafu::OptionExt;
+use snafu::ResultExt;
+use std::time::SystemTime;
+
+use tokio::time::{sleep, Duration};
+
+use crate::ec2_provider::DEFAULT_INSTANCE_COUNT;
+
+use models::{
+    constants::NAMESPACE,
+    node::{BottlerocketShadow, BottlerocketShadowState},
+};
+
+const MONITOR_SLEEP_DURATION: Duration = Duration::from_secs(30);
+
+pub type Result<T> = std::result::Result<T, monitor_error::Error>;
+
+#[async_trait]
+/// A trait providing an interface to interact with brupop recourse objects. This is provided as a trait
+/// in order to allow mocks to be used for testing purposes.
+pub trait BrupopClient: Clone + Sync + Send {
+    // fetch BottlerocketShadows and help to get node `metadata`, `spec`, and  `status`.
+    async fn fetch_shadows(&self) -> Result<ObjectList<BottlerocketShadow>>;
+    // fetch brupop resources - Controllers, Agents, Apiserver to help on determining if they are on ideal status.
+    async fn fetch_brupop_resources(&self) -> Result<ObjectList<Pod>>;
+}
+
+#[derive(Clone)]
+/// Concrete implementation of the `BrupopClient` trait. This implementation will almost
+/// certainly be used in any case that isn't a unit test.
+pub struct IntegBrupopClient {
+    k8s_client: kube::client::Client,
+}
+
+impl IntegBrupopClient {
+    pub fn new(k8s_client: kube::client::Client) -> Self {
+        IntegBrupopClient { k8s_client }
+    }
+}
+
+#[async_trait]
+impl BrupopClient for IntegBrupopClient {
+    async fn fetch_shadows(&self) -> Result<ObjectList<BottlerocketShadow>> {
+        let brss: Api<BottlerocketShadow> = Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+        let brss_object_list = brss
+            .list(&ListParams::default())
+            .await
+            .context(monitor_error::FindBrupopResources {})?;
+
+        Ok(brss_object_list)
+    }
+
+    async fn fetch_brupop_resources(&self) -> Result<ObjectList<Pod>> {
+        let pods: Api<Pod> = Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+        let pods_objectlist = pods
+            .list(&ListParams::default())
+            .await
+            .context(monitor_error::FindBrupopResources {})?;
+
+        Ok(pods_objectlist)
+    }
+}
+
+#[async_trait]
+/// A trait providing an interface to monitor brupop process.
+pub trait Monitor: Clone {
+    async fn run_monitor(&self) -> Result<()>;
+}
+
+#[derive(Clone)]
+pub struct BrupopMonitor<T: BrupopClient> {
+    integ_brupop_client: T,
+}
+
+impl<T: BrupopClient> BrupopMonitor<T> {
+    pub fn new(integ_brupop_client: T) -> Self {
+        BrupopMonitor {
+            integ_brupop_client,
+        }
+    }
+
+    // verify if Burpop resources (agent, api-server, controller) are in `running` status.
+    async fn check_resources_health(&self, pods: &ObjectList<Pod>) -> Result<bool> {
+        for pod in pods {
+            if !is_pod_running(&pod) {
+                return Ok(false);
+            }
+        }
```

Comment:
Not sure if you've played much with the Iterator API, but this might be more succinctly written as:

```rust
Ok(pods.iter().any(|pod| !is_pod_running(pod))
```

The current implementation is fine, it's just a tip. I think you might be able to return just a `bool` here instead of a `Result` though.